pkg/test/framework/components/echo/call.go

```diff
@@ -73,56 +74,127 @@ type CallOptions struct {
+		return err != nil, err
+	})
+)
+
+// ExpectNoError returns a Validator that is completed when no error occurs.
+func ExpectNoError() Validator {
+	return expectNoError
 }
 
-// WithOK returns a copy of this Validators with ValidateOK added.
-func (all Validators) WithOK() Validators {
-	return all.With(ValidateOK)
+// ExpectError returns a Validator that is completed when an error occurs.
+func ExpectError() Validator {
+	return expectError
 }
 
-// WithCount returns a copy of this Validators with validation for the given
-// expected response count.
-func (all Validators) WithCount(expectedCount int) Validators {
-	return all.With(func(responses client.ParsedResponses) error {
+// ExpectOK returns a Validator that calls CheckOK on the given responses.
+func ExpectOK() Validator {
+	return ValidatorFunc(func(resp client.ParsedResponses, err error) (bool, error) {
+		return expectNoError.Validate(resp, resp.CheckOK())
+	})
+}
+
+// ExpectCount returns a Validator that checks the number of responses against the expected.
+func ExpectCount(expectedCount int) Validator {
+	return ValidatorFunc(func(responses client.ParsedResponses, err error) (bool, error) {
 		if len(responses) != expectedCount {
-			return fmt.Errorf("unexpected number of responses: expected %d, received %d",
+			err = fmt.Errorf("unexpected number of responses: expected %d, received %d",
 				expectedCount, len(responses))
 		}
-		return nil
+		return expectNoError.Validate(responses, err)
+	})
+}
+
+// ExpectReachedClusters returns a Validator that checks that all provided clusters are reached.
+func ExpectReachedClusters(clusters resource.Clusters) Validator {
+	return ValidatorFunc(func(responses client.ParsedResponses, _ error) (bool, error) {
+		return expectNoError.Validate(responses, responses.CheckReachedClusters(clusters))
+	})
+}
+
+// ExpectHost returns a Validator that checks the responses for the given host header.
+func ExpectHost(expected string) Validator {
+	return ValidatorFunc(func(responses client.ParsedResponses, _ error) (bool, error) {
+		return expectNoError.Validate(responses, responses.CheckHost(expected))
 	})
 }
 
-// With returns a copy of this Validators with the given validator added.
-func (all Validators) With(v Validator) Validators {
-	return append(append(NewValidators(), all...), v)
+// ExpectCode returns a Validator that checks the responses for the given response code.
+func ExpectCode(expected string) Validator {
+	return ValidatorFunc(func(responses client.ParsedResponses, _ error) (bool, error) {
+		err := responses.CheckCode(expected)
+		return expectNoError.Validate(responses, err)
+	})
 }
 
-var _ Validator = ValidateOK
+// ValidatorFunc converts the given function into a Validator
+func ValidatorFunc(f func(client.ParsedResponses, error) (bool, error)) Validator {
+	return validatorFunc(f)
+}
+
+type validatorFunc func(client.ParsedResponses, error) (bool, error)
```

Comment:
I am a bit confused what `bool` in the response is. It seems like its one of those variables where its redundant since its always equivilent to `error==nil`? is there any uses for it?

Reply:
Also I think this should be a type alias not a type declaration then we do not need `ValidatorFunc` wrapper?

Reply:
I went back and forth on this a bit. Basically, this is due to the `retry` API, which returns `completed` as well as `error`.  Without `completed`, those functions that expect an error could not return it. This may be fine, not sure.

Good suggestion on the type alias ... will switch to that.

Reply:
I went through the whole code and did not find a single usage where the `bool` response was not equivalent to `error==nil` - might have missed one though. Can we remove it from the interface and just insert it as part of callInternal?

Also, why is error an input to the validator function? I have a hard time understanding how to use ValidatorFunc - it could use some comments about the expected input and output params